tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc

```diff
@@ -183,6 +198,37 @@ class HloToLhloOpConverter<mhlo::DotOp> : public BaseOpConversion<mhlo::DotOp> {
   }
 };
 
+template <typename HloOpTy>
+class HloTupleOutputToLhloOpConverter : public BaseOpConversion<HloOpTy> {
+public:
+  using BaseOpConversion<HloOpTy>::BaseOpConversion;
+  LogicalResult matchAndRewrite(
+    HloOpTy hloOp, ArrayRef<Value> operands,
+    ConversionPatternRewriter& rewriter) const final {
+    Operation* op = hloOp.getOperation();
+    SmallVector<Operation*, 4> allUsers;
+    for (auto* user : op->getUsers()) {
+      allUsers.push_back(user);
+      if (!isa<mhlo::GetTupleElementOp>(user)) {
+        return failure();
+      }
+    }
+    SmallVector<Value, 4> buffer_args(operands.begin(), operands.end());
+    if (failed(ConvertResults(op, buffer_args, rewriter))) return failure();
+    rewriter.create<mhlo::HloToLhloOp<HloOpTy>>(op->getLoc(), llvm::None,
+      buffer_args, op->getAttrs());
+    for (auto* user : allUsers) {
+      auto getElementOp = cast<mhlo::GetTupleElementOp>(user);
+      unsigned index = op->getNumOperands() + getElementOp.index();
+      getElementOp.getResult().replaceAllUsesWith(buffer_args[index]);
+      rewriter.eraseOp(user);
```

Comment:
This sequence isn't save: the replacement does not go through the `rewriter` and it won't be able to rollback or can gets its internal state confused. That can lead to crashes instead of diagnostics for example.

Let me try to rewrite this.

```suggestion
      rewriter.replaceOp(getElementOp, {buffer_args[index]});
```